modules/aruco/src/charuco.cpp

```diff
@@ -0,0 +1,362 @@
+        if (innerRect.contains(_allImgCorners.getMat().ptr<cv::Point2f>(0)[i])) {
+            filteredChessboardImgPoints.push_back(_allImgCorners.getMat().ptr<cv::Point2f>(0)[i]);
+            filteredChessboardObjPoints.push_back(board.chessboardCorners[i]);
+        }
+    }
+
+    // if none valid, return 0
+    if (filteredChessboardImgPoints.size() == 0) return 0;
+
+    // corner refinement
+    cv::Mat grey;
+    if (_image.getMat().type() == CV_8UC3)
+        cv::cvtColor(_image.getMat(), grey, cv::COLOR_BGR2GRAY);
+    else
+       _image.getMat().copyTo(grey);
+    DetectorParameters params; // use default params for corner refinement
+    cv::cornerSubPix(grey, filteredChessboardImgPoints,
+                     cvSize(params.cornerRefinementWinSize, params.cornerRefinementWinSize),
+                     cvSize(-1, -1),
+                     cvTermCriteria(CV_TERMCRIT_ITER | CV_TERMCRIT_EPS,
+                                    params.cornerRefinementMaxIterations,
+                                    params.cornerRefinementMinAccuracy));
+
+    // update refinement in input _allImgCorners
+    unsigned int idx = 0;
+    for (unsigned int i=0; i<_allImgCorners.getMat().total(); i++) {
+        if (innerRect.contains(_allImgCorners.getMat().ptr<cv::Point2f>(0)[i])) {
+            _allImgCorners.getMat().ptr<cv::Point2f>(0)[i] = filteredChessboardImgPoints[idx];
+            idx++;
+        }
+    }
+
+    // parse output
+    _selectedImgCorners.create((int)filteredChessboardImgPoints.size(), 1, CV_32FC2);
+    for (unsigned int i = 0; i < filteredChessboardImgPoints.size(); i++) {
+        _selectedImgCorners.getMat().ptr<cv::Point2f>(0)[i] = filteredChessboardImgPoints[i];
+    }
+
+    _selectedObjCorners.create((int)filteredChessboardObjPoints.size(), 1, CV_32FC3);
+    for (unsigned int i = 0; i < filteredChessboardObjPoints.size(); i++) {
+        _selectedObjCorners.getMat().ptr<cv::Point3f>(0)[i] = filteredChessboardObjPoints[i];
+    }
+
+    return filteredChessboardImgPoints.size();
+}
+
+
+
+/**
+  */
+bool estimatePoseCharucoBoard(InputArrayOfArrays _corners, InputArray _ids, InputArray _image,
+                              const CharucoBoard &board, InputArray _cameraMatrix,
+                              InputArray _distCoeffs, OutputArray _rvec, OutputArray _tvec,
+                              OutputArray _chessboardCorners) {
+
+    CV_Assert(_image.getMat().channels() == 1 || _image.getMat().channels() == 3);
+    CV_Assert(_corners.total() == _ids.total() && _ids.total() > 0);
+
+    // approximated pose estimation
+    cv::Mat approximatedRvec, approximatedTvec;
+    int detectedBoardMarkers;
+    detectedBoardMarkers = cv::aruco::estimatePoseBoard(_corners, _ids, board, _cameraMatrix,
+                                                        _distCoeffs, approximatedRvec,
+                                                        approximatedTvec);
```

Comment:
I think one global Pose for the board is bad because you have only an approximate ```_cameraMatrix``` and ``` _distCoeffs```. I think you would need to get a local homography per marker and then apply that local homography to the canonical chessboard corners to get to chessboard corners in the image. The local homography would handle distortions better. If you get one transformation per board, it is easy to get the subpixel corner accuracy wrong when the board is in extreme angle.

Also, once you have a local homography for each marker, the window size of the subpixel corner detection algorithm can be resized by the local homography. So, the window size of the subpixel corner detection depends on 2 things: (1) homography (2) ratio of sizes of aruco marker to the checkerboard square size.